### PR TITLE
Naming

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: NAMING.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+---
+
+In general Angular2 should follow Java naming conventions.
+
+
+Classes:
+  - Example: `Compiler`, `ApplicationMetadata`
+  - Camel case with first letter upper-case
+  - In general preffer single words. (This is so that when appending `Proto` or `Factory` the class 
+    is still reasonable to worke with.)
```
Comment (vicb) [inline NAMING.md]: work with

File: NAMING.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  - Should not end with `Impl` or any other word which describes a specefic implemenation of an
```
Comment (vicb) [inline NAMING.md]: "specific"

Consider the following case: we have two different implementations of the class (JS & Dart), what would you recommend ?

Currently we have `XHRImpl extends XHR` - I ran into the same question for `NgZone` in a pending PR

File: NAMING.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  
+Methods and functions:
+  - Example: `bootstrap`, `ngProbe`
+  - Should be camel case with first lower case
+
+
+Constants
+  - Example: `CORE_DIRECTIVES`
+  - Should be all upercase with snake_case
```
Comment (vicb) [inline NAMING.md]: SNAKE_CASE

**Comment by vicb** [comment]
May be add a note about private member requiring a "_" prefix ?

**Comment by vicb** [comment]
Also what about renaming this to "STYLE_GUIDE.md" so that we can add more things later (à la https://google.github.io/styleguide/javascriptguide.xml)

**Comment by btford** [comment]
Would be nice to include justification for each item.

File: NAMING.md (hunk 3 above)
Comment (gkalpak) [inline NAMING.md]: _upercase_  -->  _uppercase_

File: NAMING.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  - Camel case with first letter upper-case
+  - In general preffer single words. (This is so that when appending `Proto` or `Factory` the class 
+    is still reasonable to worke with.)
+  - Should not end with `Impl` or any other word which describes a specefic implemenation of an 
+    interface.
+  
+  
+Interfaces:
+  - Follow the same rules as Classes 
+  - Should not have `I` or `Interface` in the name or any other way of identifing it as an interaface.
```
Comment (gkalpak) [inline NAMING.md]: _interaface_  -->  _interface_

File: NAMING.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+Naming Conventions in Angular2
+---
+
+In general Angular2 should follow Java naming conventions.
+
+
+Classes:
+  - Example: `Compiler`, `ApplicationMetadata`
+  - Camel case with first letter upper-case
+  - In general preffer single words. (This is so that when appending `Proto` or `Factory` the class
```
Comment (gkalpak) [inline NAMING.md]: _preffer_  --> _prefer_

File: NAMING.md (hunk 2 above)
Comment (mhevery) [inline NAMING.md]: Shouldn't it be `DartXhr` and `TypeScriptXHR` then?

If we don't have Dart, then `TypeScriptXHR` would not make sense, which would argue that the separation is not needed.

**Comment by mhevery** [comment]
@vicb, added your comments. 
@btford +1 on justification, but many of them are just convention. Why is method `fooBar`, class `FooBar` and const `FOO_BAR`. There really is no real justification, other then, it is what people are use to.

File: NAMING.md (hunk 2 above)
Comment (vicb) [inline NAMING.md]: Both the implementation have to have the same name, otherwise `bind(Xhr).toClass(XhrImpl)` would not be possible.

Comment (mhevery) [inline NAMING.md]: Then I will argue that we should replace bind(Xhr).toClass(XhrImpl) with
just Xhr. If there is exactly one implementation of something then the
interface is meaningless.

Comment (vicb) [inline NAMING.md]: We actually have `XhrImpl (JS)`, `XhrImpl (Dart)` and `XhrMock (Test)`.

IMO having `XhrMock` extending a concrete implementation would feel both ugly and hacky which is why we have an abstract base class.

**Comment by gkalpak** [comment]
It might be too specific, but I see inconsistencies often arise with acronyms (XHR, URL etc).

I personally prefer to only capitalize the first letter (as in `XhrService`, `urlThingy`, `customUrl`).
It is good to have an explicit convention for such cases (one way or another), to avoid inconsistencies.
(Besides the unintuitiveness of `FancyFoo` vs `FancyBAR` that will eventually come up, I am more concenred about _eyes-hurting_ "contraptions" such as `rAF` vs `RAFProvider` vs `rAFMutex` vs `rafWaitQueue` (as seen in ng1).)

**Comment by gkalpak** [comment]
It's happening already (unless I am missing something):
Does anyone else find it _unintuitive_ that there is: `DOM` vs `DomAdapter` and `DomAttrAST` (note `Dom...` vs `...AST`) ?

**Comment by vicb** [comment]
@gkalpak agree, would be nice to have a convention here (no presonal prefs here)

**Comment by mhevery** [comment]
@gkalpak Agreed, but it is private API, and so it will not effect our users and as such not a high priority.

**Comment by gkalpak** [comment]
FWIW, it is also happening with animate's ` RAF`, which I think is public (but maybe I'm wrong).

It's definitely not high priority to change existing or private names, but it might be a good idea to have a guideline for newly added stuff.

**Pull request closed without merging**